nemo/collections/nlp/modules/common/megatron/transformer.py

```diff
@@ -510,7 +510,9 @@ def forward(
             # adjust the key rotary positional embedding
             if rotary_pos_emb is not None:
                 q_pos_emb, k_pos_emb = rotary_pos_emb
-                k_pos_emb = k_pos_emb[:, :, :end, :]
+                if not set_inference_key_value_memory:
+                    q_pos_emb = q_pos_emb[end - 1 : end]
```

Comment:
Leaving comment here to discuss this line

Reply:
this is used in inference. We compute one token at a time. this is used to select the correct positional embedding.